### PR TITLE
Fix for deleted dbCheckpoints when scale with --restart executed

### PR DESCRIPTION
**Pull request by itsmeakashgoyal**
Fix for deleted dbCheckpoints when scale command executed with restart flag

* **Problem Overview**  
  All created db checkpoints deleted from file system when scale with --restart executed.
* **Testing Done**  
  Apollo test is added and same has been tested.
  Also performed maestro multiday run which results no impact on performance.

### CODE REVIEW

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 1]
```diff
@@ -115,14 +116,62 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
+    try {
+      if (_fs::exists(path)) {
+        for (const auto& entry : _fs::directory_iterator(checkpointDir)) {
+          const auto filenameStr = entry.path().filename().string();
+          if (_fs::is_directory(entry)) {
+            allDBCheckpoints.push_back(entry);
+            const CheckpointId& checkPointId = std::stoi(filenameStr);
+            const auto blockId = checkPointId;
+
+            auto p = path / filenameStr;
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:136]: why not use entry.path()?

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 2]
```diff
@@ -115,14 +116,62 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
+              lastCreatedCheckpointMetadata_.emplace(DbCheckpointMetadata::DbCheckPointDescriptor{
+                  checkPointId, creationTime, blockId, lastCheckpointSeqNum_});
+              dbCheckptMetadata_.dbCheckPoints_.insert({checkPointId, lastCreatedCheckpointMetadata_.value()});
+            }
+            numOfDbCheckpointsCreated_++;
+          }
+        }
+
+        // extract the last created BlockId from all created dbCheckpoints.
+        auto lastCreatedDbCheckpoint =
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:151]: would it be simple to just use the highest checkpoint id seen in the checkpoint folder in the for loop above. 
Something like, 
lastCreatedDbCheckpoint = std::max(lastCreatedDbCheckpoint, stoi(checkpoint_id from directory iterator));

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 3]
```diff
@@ -115,14 +116,62 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
+                              [](const _fs::directory_entry& f1, const _fs::directory_entry& f2) {
+                                return _fs::last_write_time(f1.path()) < _fs::last_write_time(f2.path());
+                              });
+
+        lastDbCheckpointBlockId_.Get().Set(std::stoi(lastCreatedDbCheckpoint.path().filename()));
+        metrics_.UpdateAggregator();
+      }
+      updateMetrics();
+    } catch (std::exception& e) {
+      LOG_WARN(getLogger(), "Failed to find any checkpoints: " << e.what());
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:163]: Warning message could be: failed to update checkpoints metadata from checkpoint directory

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 4]
```diff
@@ -115,14 +116,62 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
-  loadCheckpointDataFromPersistence();
-  // if there is a checkpoint created in database but entry corresponding to that is not found in persistence
-  // then probably, its a partially created checkpoint and we want to remove it
-  auto listOfCheckpointsCreated = dbClient_->getListOfCreatedCheckpoints();
-  for (auto& cp : listOfCheckpointsCreated) {
-    if (dbCheckptMetadata_.dbCheckPoints_.find(cp) == dbCheckptMetadata_.dbCheckPoints_.end()) {
-      removeCheckpoint(cp);
+  // In case of scale with restart we have to update dbcheckpoint metadata from file system
+  // if any dbcheckpoints are present in the file system.
+  if (lastCheckpointSeqNum_ == 0 && EpochManager::instance().getSelfEpochNumber() > 0) {
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:121]: to be more precise, we can also check if we have scale command logged and its epoch number is one less than the current epoch number. (may be in future PR)

**Review by nkumar04 — APPROVED** [review]
LGTM!
Thanks!

**Comment by nityavasudevan** [comment]
Please update with maestro run  results in test results

**Comment by itsmeakashgoyal** [comment]
> Please update with maestro run results in test results

Hi @nityavasudevan I have performed the perf run with multiday activity and the overall TPS is more than 630.

**Review by arc-vmware — APPROVED** [review]
LGTM!!!

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 5]
```diff
@@ -115,14 +115,56 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
+  // In case when metadata erased, we have to update dbcheckpoint metadata from file system
+  // if any dbcheckpoints are present in the file system.
+  if (isMetadataErased_) {
+    // update metadata for dbcheckpoints from file system
+    const auto& checkpointDir = dbClient_->getCheckpointPath();
+    _fs::path path(checkpointDir);
+    std::vector<_fs::directory_entry> allDBCheckpoints;
+
+    try {
+      if (_fs::exists(path)) {
```
Comment (cloudnoize) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:127]: This code should have been implemented in a separate method with a unit test

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 6]
```diff
@@ -115,14 +115,56 @@ void DbCheckpointManager::loadCheckpointDataFromPersistence() {
...
+            allDBCheckpoints.push_back(entry);
+            const CheckpointId& checkPointId = std::stoi(filenameStr);
+            const auto blockId = checkPointId;
+
+            _fs::file_time_type ftime = _fs::last_write_time(entry.path());
+            std::time_t cftime = decltype(ftime)::clock::to_time_t(ftime);
+            const std::chrono::seconds creationTime = std::chrono::seconds(cftime);
+            {
+              std::scoped_lock lock(lockLastDbCheckpointDesc_);
+              lastCreatedCheckpointMetadata_.emplace(DbCheckpointMetadata::DbCheckPointDescriptor{
```
Comment (cloudnoize) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:141]: If multiple checkpoint directories exist, why do you assume that ```lastCreatedCheckpointMetadata_``` will contain the last checkpoint?
From the directory_iterator doc ```The iteration order is unspecified```

**Pull request merged**